### PR TITLE
Improve cagg watermark caching

### LINKED ISSUE #2826: Real-time aggregation sometimes slow
The cagg_watermark function used to find the boundary where to join raw and materialized data is marked STABLE, but the planner still executes it multiple times (sometimes once per row) for the same input argument in a single query. This makes real-time aggregation slow.

This was fixed by caching the watermark in `fn_extra`, part of the function state, but this also seems to not work in some cases.

Another solution might be to cache the watermark as a static variable for the duration of a transaction.

### PR DESCRIPTION
**Pull request by erimatnor**
The internal function `cagg_watermark` returns the last/max bucket in
a continuous aggregate and is used to get the point where to union the
raw and materialized data when using real-time aggregation.

Since the function is marked `STABLE` the planner should be able to
constify the function so that it need not be called repeatedly during
execution. However, this optimization is not guaranteed and the
function might be constified many times during planning if it occurs,
e.g., as an index scan condition on many chunks. This leads to slow
queries due to repeated calls to the function.

Previously, the watermark was cached in the function's `fn_extra`
state, but this state doesn't survive all repeated calls of the
function and doesn't help if the function occurs many times in the
same plan.

To make the performance more reliable, the watermark is now cached in
a global variable that is cleared at transaction end, when the input
argument (materialized hypertable queried) changes, or a new command
is executed. This should guarantee that the watermark only needs to be
looked up once per query.

Fixes #2826

### CODE REVIEW

**Review by gayyappan — APPROVED** [review]

**Review by pmwkaa — APPROVED** [review]

File: src/continuous_agg.c [hunk 1]
```diff
@@ -935,13 +1009,12 @@ ts_continuous_agg_watermark(PG_FUNCTION_ARGS)
...
-		 * the documentation isn't clear on this. Check that the hypertable ID
-		 * matches across repeated calls just to be safe. */
-		if (watermark->hyper_id == hyper_id)
+		/* Check that the hypertable ID matches across repeated calls. Reset
+		 * the cached watermark if the ID changes. */
+		if (watermark_valid(watermark, hyper_id))
 			PG_RETURN_INT64(watermark->value);
 
-		pfree(watermark);
+		MemoryContextDelete(watermark->mctx);
```
Comment (pmwkaa) [inline src/continuous_agg.c:1017]: This one will also be called when `top_mctx` gets deleted as well, right?

**Review by pmwkaa — APPROVED** [review]
looks good

File: src/continuous_agg.c (hunk 1 above)
Comment (erimatnor) [reply src/continuous_agg.c:1017]: Yes. If you destroy a parent, then all its children will be deleted.

**Review by pmwkaa — APPROVED** [review]

File: tsl/test/expected/continuous_aggs_query-11.out [hunk 2]
```diff
@@ -721,3 +721,143 @@ select * from mat_m1 order by timec desc, location;
...
+SELECT chunk_name, range_start, range_end
+FROM timescaledb_information.chunks
+WHERE hypertable_name = :'mat_name';
+    chunk_name    |         range_start          |          range_end           
+------------------+------------------------------+------------------------------
+ _hyper_2_3_chunk | Wed Nov 29 16:00:00 2017 PST | Wed Feb 07 16:00:00 2018 PST
+(1 row)
+
+-- The watermark should be updated to reflect the dropped data (i.e.,
+-- the cache should be reset)
```
Comment (gayyappan) [inline tsl/test/expected/continuous_aggs_query-11.out:820]: useful to add select max(time) from mat_m1 here. I had to go back to the data to identify this.
and select max(time) from mat_m2.

File: tsl/test/expected/continuous_aggs_query-11.out [hunk 3]
```diff
@@ -721,3 +721,143 @@ select * from mat_m1 order by timec desc, location;
...
+ NYC      | Wed Oct 31 17:00:00 2018 PDT |  325 |  200 |     30 |    50 |      85 |      45
+ NYC      | Mon Jan 01 16:00:00 2018 PST |   65 |   45 |     45 |    45 |      65 |      65
+ SFO      | Sun Dec 02 16:00:00 2018 PST |  120 |   90 |     45 |    45 |      65 |      55
+ SFO      | Mon Jan 01 16:00:00 2018 PST |   65 |   45 |     45 |    45 |      65 |      65
+ SFO      | Sun Dec 31 16:00:00 2017 PST |   55 |   45 |     45 |    45 |      55 |      55
+ por      | Mon Jan 01 16:00:00 2018 PST |  100 |  100 |    100 |   100 |     100 |     100
+(9 rows)
+
+ALTER MATERIALIZED VIEW mat_m1 SET (timescaledb.materialized_only=true);
+SELECT m1.location, m1.timec, sumt, sumh, firsth, lasth, maxtemp, mintemp
```
Comment (gayyappan) [inline tsl/test/expected/continuous_aggs_query-11.out:849]: SELECT m2.location, m2.timec, sumt, sumh, firsth, lasth, maxtem, mintemp
FROM mat_m1 RIGHT JOIN mat_m2 ON ....

This would also verify/highlight the difference between the 2  caggs.

**Review by gayyappan — APPROVED** [review]
Changes look good. Made a few suggestions to add to the test.

File: tsl/test/expected/continuous_aggs_query-11.out (hunk 2 above)
Comment (erimatnor) [reply tsl/test/expected/continuous_aggs_query-11.out:838]: Done. Also comparing to invalidation threshold.

File: tsl/test/expected/continuous_aggs_query-11.out (hunk 3 above)
Comment (erimatnor) [reply tsl/test/expected/continuous_aggs_query-11.out:884]: Done.

File: tsl/test/sql/continuous_aggs_query.sql.in [hunk 4]
```diff
@@ -26,19 +26,19 @@ CREATE TABLE conditions (
...
+insert into conditions values ( '2018-01-02 09:10:00-08', 'NYC', 65, 45);
+insert into conditions values ( '2018-11-01 09:20:00-08', 'NYC', 45, 30);
+insert into conditions values ( '2018-11-01 10:40:00-08', 'NYC', 55, 35);
+insert into conditions values ( '2018-11-01 11:50:00-08', 'NYC', 65, 40);
+insert into conditions values ( '2018-11-01 12:10:00-08', 'NYC', 75, 45);
+insert into conditions values ( '2018-11-01 13:10:00-08', 'NYC', 85, 50);
+insert into conditions values ( '2018-11-02 09:20:00-08', 'NYC', 10, 10);
+insert into conditions values ( '2018-11-02 10:30:00-08', 'NYC', 20, 15);
+insert into conditions values ( '2018-11-02 11:40:00-08', 'NYC', null, null);
+insert into conditions values ( '2018-11-03 09:50:00-08', 'NYC', null, null);
```
Comment (mkindahl) [inline tsl/test/sql/continuous_aggs_query.sql.in:41]: Nit: Not that it really makes a difference, but why this change? It does not seem warranted.

File: tsl/test/sql/continuous_aggs_query.sql.in [hunk 5]
```diff
@@ -187,3 +187,107 @@ set timescaledb.enable_cagg_reorder_groupby = false;
...
+insert into conditions values ( '2018-12-02 20:30:00-08', 'NYC', 65, 45);
+insert into conditions values ( '2018-12-02 21:50:00-08', 'NYC', 45, 30);
+
+-- Test join of two caggs. Joining two caggs will force the cache to
+-- reset every time the watermark function is invoked on a different
+-- cagg in the same query.
+SELECT mat_hypertable_id AS mat_id, raw_hypertable_id AS raw_id, schema_name AS mat_schema, table_name AS mat_name, format('%I.%I', schema_name, table_name) AS mat_table
+FROM _timescaledb_catalog.continuous_agg ca, _timescaledb_catalog.hypertable h
+WHERE user_view_name='mat_m1'
+AND h.id = ca.mat_hypertable_id \gset
```
Comment (mkindahl) [inline tsl/test/sql/continuous_aggs_query.sql.in:212]: Nit: Very long line, might be better to split each field on a separate line.

**Review by mkindahl — APPROVED** [review]
Good job on the tests, I think it covers the situations. Would be good if there were a way to check that the cached variable is really gone after the transaction (e.g., a separate test function that calls `watermark_valid` or so), but that is a stretch.

File: tsl/test/sql/continuous_aggs_query.sql.in (hunk 4 above)
Comment (erimatnor) [reply tsl/test/sql/continuous_aggs_query.sql.in:41]: I didn't want to align on hours since bucketing typically falls on some hour of the day. This makes it easier to see that the value returned by the watermark (max-derived value) is a bucket and not an inserted value. I can revert if you feel it is important to keep the same.

**Comment by erimatnor** [comment]
> Good job on the tests, I think it covers the situations. Would be good if there were a way to check that the cached variable is really gone after the transaction (e.g., a separate test function that calls `watermark_valid` or so), but that is a stretch.

I thought about that, but it requires additional boilerplate code to essentially test that PostgreSQL memory context callbacks work the way they should. I did manually test this, FWIW.

I think we mostly care about the fact that we return the "right" watermark, which is covered  by tests, and less with what happens outside a transaction. The only concern there would be a memory leak, I guess; but then we are testing memory contexts.

File: src/continuous_agg.c [hunk 6]
```diff
@@ -897,9 +898,88 @@ ts_continuous_agg_find_integer_now_func_by_materialization_id(int32 mat_htid)
...
+ * hypertable ID) changes, or when a new command is executed. One could
+ * potentially create a hashtable of watermarks keyed on materialized
+ * hypertable ID, but this is left as a future optimization since it doesn't
+ * seem to be common case that multiple continuous aggregates exist in the
+ * same query. Besides, the planner can constify calls to the watermark
+ * function during planning since the function is STABLE. Therefore, this is
+ * only a fallback if the planner needs to constify it many times (e.g., if
+ * used as an index condition on many chunks).
+ */
+static Watermark *watermark = NULL;
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/continuous_agg.c:919]: I changed the name of the local scope variable
